_docs/developer/testing/cypress.md

```diff
@@ -0,0 +1,187 @@
+
+The `support/index.js` file instructs Cypress which files to include with every test, you most likely won't need to touch this file. 
+
+The `plugins/index.js` allows you to extend Cypress's default behavior. You can read more about them [here](https://docs.cypress.io/guides/tooling/plugins-guide.html#Use-Cases). 
+
+The top level `cypress.json` is the global configuration used.
+
+### Test Structure
+
+Cypress borrow's Mocha's Behavior-driven development (BDD) [test interface](https://mochajs.org/#bdd) to organize tests. 
+
+You will typically use `describes`, `context`, and `it` to organize tests. `describes` and `context` organize tests into groups, either or can be used they both act the same. Here is an example:
+
+```js
+describe('foo', () => {
+	context('bar', () => {
+		it('baz', () => {
+			expect(true).to.equal(true);
+		});
+	});
+});
+```
+
+This will produce the following structure in the test runner:
+
+![Cypress organization in the test runner](/images/cypress-structure.png)
+
+You can also use the `afterEach` and `beforeEach` function calls to perform setup and teardown for each test. 
+Be sure to read Cypress's best-practices for using these functions [here](https://docs.cypress.io/guides/references/best-practices.html#Using-after-or-afterEach-hooks).
+
+### Workflow 
+
+A test will typically:
+
+1. Navigate to a page
+2. Find an element
+3. *(Optionally) Interact with it*
+4. *(Optionally) Make an assertion*
+
+This translates to cypress functions in the form of:
+
+1. `cy.visit('/your_page/')`
+2. `cy.get('your_element')`
+3. [Interaction](https://docs.cypress.io/guides/core-concepts/interacting-with-elements.html#Actionability)
+4. [Assertions](https://docs.cypress.io/guides/references/assertions.html#Chai)
+
+Cypress allows you to select elements through [JQuery Selectors](https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Cypress-is-Like-jQuery). 
+
+You can also use the "selector playground" in the test runner which allows you to highlight part of the screen
+and let Cypress generate a selector for you. *Note:* for deeply nested HTML elements you will most likely have to create the selector yourself.
+
+![Cypress selector-playground example](/images/selector-playground.png)
+
+You can then use the `cy.get()` function to attempt to find an element on the page. If your test only requires an element to exist on the page, using a `cy.get()` is a strong enough check. 
+
+If an element is not immediately ready Cypress will automatically retry for 4000ms by default, this can be useful for elements that dynamically render or preventing flaky-ness.
+
+**NOTE**
+
+Most Cypress functions are asynchronous and yield a result instead of returning it, this requires you to chain
+together actions.
+
+```js
+cy.get('my_element')
+.then((result) => {
+	//do something with the element we found
+});
```

Comment:
Can we express this example using the async / await syntax? We should be encouraging it to avoid having huge promise chains if we can help it.

Reply:
So cypress specifically doesn't allow async/await, since it clashes with their API design / doesn't work.

From https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Commands-Are-Asynchronous

>  Why can’t I use async / await?
>
> If you’re a modern JS programmer you might hear “asynchronous” and think: why can’t I just use async/await instead of learning some proprietary API?
>
> Cypress’s APIs are built very differently from what you’re likely used to: but these design patterns are incredibly intentional. We’ll go into more detail later in this guide.

This is probably due to Cypress implementing a version of something that behaves similar to promises but not exactly 
https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Commands-Are-Not-Promises

Reply:
Can we add a note and link to their docs on this?